Threading `map()` implementation using thread-local parser copies

levitsky opened the pull request:
This implements a variant of the fist approach in #174. Namely, a shallow copy of the reader object is created <s>and put in a `threading.local`,</s> so every working thread uses its own copy. The copies all use the same offset index, so it is not recreated, however the current implementation of `_file_obj` ensures that a new file object is created upon copying, which allows all copies to read the same file via multiple file descriptors.

<s>This implementation uses `concurrent.futures.ThreadPoolExecutor` to do the actual mapping. Because of this, it is hard to close the thread-local readers explicitly, so it is delegated to `gc`.</s>

Performance-wise, at least in the trivial case without a `target` function, this implementation takes about the same time as the `multiprocessing`-based `map()` call on the same Python version (actually even a bit slower). Tested on `3.13.1t` and `3.14.2t` with `PYTHON_GIL=0`, the single-threaded performance is significantly worse than the normal Python's, so for now the best way in most cases is still to use the default `map()`. However, in cases when large data needs to be shared and IPC becomes a bottleneck, or when serializing is an issue, using threading can be advantageous.

Counter-intuitively, with MzIdentML, where additional I/O is done to `retrieve_refs`, the performance with threading `map` is actually better in some cases than that achieved on normal Python with processes.

mobiusklein commented on pyteomics/auxiliary/file_helpers.py:
```diff
@@ -1121,26 +1134,206 @@ def iterate():
+                    if key is _SENTINEL:
+                        # Do not put the sentinel back; each worker consumes one
+                        break
+                    worker_func(key, output_queue, local_reader)
```
What is the value of not simply inlining `worker_func` here? This adds an extra function call overhead for no perceivable benefit as `worker_func` is a local function and not something that can be overridden.

mobiusklein commented on pyteomics/auxiliary/file_helpers.py:
```diff
@@ -1121,26 +1134,130 @@ def iterate():
+class TaskMappingMixin(MultiProcessingTaskMappingMixin, ThreadingTaskMappingMixin):
+    pmap = MultiProcessingTaskMappingMixin.map
+    tmap = ThreadingTaskMappingMixin.map
+
```
The way MRO works this will become an issue if the two execution models share methods by name than `map` and whatever they have in common from `BaseTaskMappingMixin`

mobiusklein commented:
I spent some time reading about free threading Python and C extensions since `lxml` is pretty important for parsing. I gather the `PYTHON_GIL=0` is there to prevent `lxml` and other incompatible C extensions from re-enabling the GIL?

As to mzIdentML being faster with threads, I think that's due to more deeply nested objects in the retrieve-refs laden output not needing to be serialized/deserialized compared to processes. Of course, it might depend upon the `target` function and test data. I wrote a little IPC compression shim for when I was processing a lot of raw MS data in parallel with a similar setup since the profiler highlighted the serialization was eating a lot of the time.

levitsky commented:
The test failure was not an implementation problem, but a test bug. The sorting key was not unique so the item order was not guaranteed.

Thank you once again. I believe I addressed your comments in the previous commits. If you have any other feedback it is more than welcome, otherwise I can merge.

levitsky commented:
Latest test failures are unrelated.

Two items to complete in this branch:

- [ ] update documentation
- [ ] update `Iterfind.map`

levitsky commented:
I tried adding a test to check that `IndexedIterfind.map` works with `method='t'` and it fails:

```python
ERROR: test_iterfind_map (test_mzid.MzidTest.test_iterfind_map) (method='t')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/levitsky/pyteomics/tests/test_mzid.py", line 61, in test_iterfind_map
    sum(1 for _ in r.iterfind("SpectrumIdentificationResult").map(method=method)))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/levitsky/pyteomics/tests/test_mzid.py", line 61, in <genexpr>
    sum(1 for _ in r.iterfind("SpectrumIdentificationResult").map(method=method)))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/levitsky/pyteomics/pyteomics/auxiliary/file_helpers.py", line 1277, in iterate
    raise result
  File "/home/levitsky/pyteomics/pyteomics/auxiliary/file_helpers.py", line 1238, in thread_worker
    item = local_reader[key]
           ~~~~~~~~~~~~^^^^^
  File "/home/levitsky/pyteomics/pyteomics/xml.py", line 1382, in __getitem__
    return self._get_by_index(i)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/levitsky/pyteomics/pyteomics/xml.py", line 1433, in _get_by_index
    key = index.from_index(idx)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/home/levitsky/pyteomics/pyteomics/auxiliary/file_helpers.py", line 705, in from_index
    return items[index][0]
           ~~~~~^^^^^^^
TypeError: tuple indices must be integers or slices, not str
```

The `IndexedIterfind`'s `__getitem__` (inherited from `Iterfind`) assumes that the lookup key (if not a slice) is an integer, while its `_task_map_iterator` iterates over element IDs stored in the `OffsetIndex`, and that's what is used to retrieve items in

```python 
item = local_reader[key]
```

I think that the current multiprocessing implementation avoids this because it copies the entire parser into the processes, so the lookup happens on the parser and not on the `IndexedIterfind` object. This is because `IndexedIterfind._get_reader_for_worker_spec` returns `self.parser` so that is instantiated in `FileReadingProcess`.

I assume we should fix `Iterfind.__getitem__` to work with keys of its own index.
